plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java

```diff
@@ -978,12 +992,35 @@ protected WriteMapping legacyToWriteMapping(@SuppressWarnings("unused") Connecto
                 .orElseGet(Function::identity);
     }
 
+    protected Function<String, String> tryApplyTopN(Optional<List<SortItem>> sortItems, OptionalLong limit)
+    {
+        if (sortItems.isEmpty() || limit.isEmpty()) {
+            return Function.identity();
+        }
+
+        return orderByFunction()
+                .map(orderByFunction -> (Function<String, String>) sql -> orderByFunction.apply(sql, sortItems.get()))
+                .orElseGet(Function::identity)
+                .andThen(tryApplyLimit(limit));
+    }
+
     @Override
     public boolean supportsLimit()
     {
         return limitFunction().isPresent();
     }
 
+    @Override
+    public boolean supportsTopNPushDown()
+    {
+        return limitFunction().isPresent() && orderByFunction().isPresent();
+    }
+
+    protected Optional<BiFunction<String, List<SortItem>, String>> orderByFunction()
+    {
+        return Optional.empty();
```

Comment:
Once `supportsTopNPushDown` is removed, this can be removed as well -- another extension point is likely not needed, as `tryApplyTopN` already serves as such